snap/info_snap_yaml.go

```diff
@@ -21,18 +21,198 @@ package snap
+	for name, data := range y.Slots {
+		iface, attrs, err := convertToSlotOrPlugData("slot", name, data)
+		if err != nil {
+			return nil, err
+		}
+		snap.Slots[name] = &SlotInfo{
+			Snap:      snap,
+			Name:      name,
+			Interface: iface,
+			Attrs:     attrs,
+			Apps:      make(map[string]*AppInfo),
+		}
+	}
+	// Collect app-level implicit definitions of plugs and slots
+	for _, yApp := range y.Apps {
+		for _, plugName := range yApp.PlugNames {
+			if _, ok := snap.Plugs[plugName]; !ok {
+				snap.Plugs[plugName] = &PlugInfo{
+					Snap:      snap,
+					Name:      plugName,
+					Interface: plugName,
+					Apps:      make(map[string]*AppInfo),
+				}
+			}
+		}
+		for _, slotName := range yApp.SlotNames {
+			if _, ok := snap.Slots[slotName]; !ok {
+				snap.Slots[slotName] = &SlotInfo{
+					Snap:      snap,
+					Name:      slotName,
+					Interface: slotName,
+					Apps:      make(map[string]*AppInfo),
+				}
+			}
+		}
+	}
+	// Collect definitions of apps
+	for appName, yApp := range y.Apps {
+		snap.Apps[appName] = &AppInfo{
+			Snap:    snap,
+			Name:    appName,
+			Command: yApp.Command,
+			Plugs:   make(map[string]*PlugInfo),
+			Slots:   make(map[string]*SlotInfo),
+		}
+	}
+	// Remember which plugs and slots are app-bound
+	appBoundPlug := make(map[string]bool)
+	appBoundSlot := make(map[string]bool)
+	// Bind app-bound plugs and slots to their respective apps
+	for appName, app := range snap.Apps {
+		for _, plugName := range y.Apps[appName].PlugNames {
+			appBoundPlug[plugName] = true
+			plug := snap.Plugs[plugName]
+			app.Plugs[plugName] = plug
+			plug.Apps[appName] = app
+		}
+		for _, slotName := range y.Apps[appName].SlotNames {
+			appBoundSlot[slotName] = true
+			slot := snap.Slots[slotName]
+			app.Slots[slotName] = slot
+			slot.Apps[appName] = app
+		}
+	}
+	// Bind global plugs and slots to all apps
+	for plugName, plug := range snap.Plugs {
+		if !appBoundPlug[plugName] {
+			for appName, app := range snap.Apps {
+				app.Plugs[plugName] = plug
+				plug.Apps[appName] = app
+			}
+		}
+	}
+	for slotName, slot := range snap.Slots {
+		if !appBoundPlug[slotName] {
+			for appName, app := range snap.Apps {
+				app.Slots[slotName] = slot
+				slot.Apps[appName] = app
+			}
+		}
+	}
 	// FIXME: validation of the fields
+	return snap, nil
+}
 
-	return &s, nil
+func convertToSlotOrPlugData(plugOrSlot, name string, data interface{}) (iface string, attrs map[string]interface{}, err error) {
+	switch data.(type) {
+	case map[interface{}]interface{}:
+		for keyData, valueData := range data.(map[interface{}]interface{}) {
+			key, ok := keyData.(string)
+			if !ok {
+				return "", nil, fmt.Errorf("%s %q has attribute that is not a string (found %T)",
+					plugOrSlot, name, keyData)
+			}
+			if strings.HasPrefix(key, "$") {
+				return "", nil, fmt.Errorf("%s %q uses reserved attribute %q", plugOrSlot, name, key)
+			}
+			// XXX: perhaps we could special-case "label" the same way?
```

Comment:
The spec didn't mention it at all in the snap.yaml part. I've implemented something very simple for label support. Have a look please. What worries me is that the trivial cases (abbreviated) will be seductive for people and we won't see label used very often. Perhaps we could have some nice default labels for certain interfaces? (eg. network, network-bind, etc).